### PR TITLE
Adding trefoil controller

### PR DESCRIPTION
**Pull request by quarkytale**
Trefoil controller for publishing IMU, Magnetometer, Sea pressure and other constants from TFRecord.msg

### CODE REVIEW

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+      }
+    };
+  this->dataPtr->thread_executor_spin_ = std::thread(spin);
+
+  RCLCPP_INFO_STREAM(
+    this->dataPtr->rosnode_->get_logger(),
+    "[ROS 2 Trefoil Controller] Setting up controller for [" << model.Name(_ecm) << "]");
+
+  // IMU Sensor
+  this->dataPtr->imu_cb_ = [this](const gz::msgs::IMU & _imu)
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:159]: need a similar callback for the magnetometer

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    tf_record_.imu.angular_velocity.y = _imu.angular_velocity().y();
+    tf_record_.imu.angular_velocity.z = _imu.angular_velocity().z();
+    tf_record_.imu.linear_acceleration.x = _imu.linear_acceleration().x();
+    tf_record_.imu.linear_acceleration.y = _imu.linear_acceleration().y();
+    tf_record_.imu.linear_acceleration.z = _imu.linear_acceleration().z();
+  }
+
+  bool data_valid() const
+  {
+    return imu_data_valid_;
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:80]: `&& mag_data_valid_`

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+  std::mutex data_mutex_, next_access_mutex_, low_prio_mutex_;
+  std::atomic<bool> imu_data_valid_{false};
+
+  std::thread thread_executor_spin_, thread_publish_;
+  std::atomic<bool> stop_{false}, paused_{true};
+
+  int16_t seq_num{0};
+
+  void set_tf_record_imu(const gz::msgs::IMU & _imu)
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:60]: need a similar `set_tf_record_mag`

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+  rclcpp::Publisher<buoy_interfaces::msg::TFRecord>::SharedPtr tf_pub_{nullptr};
+  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub_{nullptr};
+
+  std::function<void(const gz::msgs::IMU &)> imu_cb_;
+  double pub_rate_hz_{10.0};
+  std::unique_ptr<rclcpp::Rate> pub_rate_{nullptr};
+  buoy_interfaces::msg::TFRecord tf_record_;
+
+  std::mutex data_mutex_, next_access_mutex_, low_prio_mutex_;
+  std::atomic<bool> imu_data_valid_{false};
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:53]: `, mag_data_valid{false}`

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+{
+  rclcpp::Node::SharedPtr rosnode_{nullptr};
+  rclcpp::executors::MultiThreadedExecutor::SharedPtr executor_;
+
+  gz::sim::Entity entity_;
+  gz::transport::Node node_;
+  std::chrono::steady_clock::duration current_time_;
+
+  rclcpp::Publisher<buoy_interfaces::msg::TFRecord>::SharedPtr tf_pub_{nullptr};
+  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub_{nullptr};
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:45]: maybe also publish the mag data straight to ros2 like the imu

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+      // low prio data access
+      std::unique_lock low(this->dataPtr->low_prio_mutex_);
+      std::unique_lock next(this->dataPtr->next_access_mutex_);
+      std::unique_lock data(this->dataPtr->data_mutex_);
+      next.unlock();
+      this->dataPtr->set_tf_record_imu(_imu);
+      this->dataPtr->imu_data_valid_ = true;
+      data.unlock();
+    };
+  if (!this->dataPtr->node_.Subscribe("/Trefoil_link/trefoil_imu", this->dataPtr->imu_cb_)) {
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:170]: need a subscriber to `"/Trefoil_link/trefoil_mag"`

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+  }
+
+  // Publisher
+  std::string tf_topic = _sdf->Get<std::string>("tf_topic", "trefoil_data").first;
+  this->dataPtr->tf_pub_ = \
+    this->dataPtr->rosnode_->create_publisher<buoy_interfaces::msg::TFRecord>(tf_topic, 10);
+
+  std::string imu_topic = _sdf->Get<std::string>("imu_topic", "trefoil_imu").first;
+  this->dataPtr->imu_pub_ = \
+    this->dataPtr->rosnode_->create_publisher<sensor_msgs::msg::Imu>(imu_topic, 10);
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:182]: maybe also publish the mag data straight to ros2 like the imu

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+  this->dataPtr->pub_rate_hz_ = \
+    _sdf->Get<double>("publish_rate", this->dataPtr->pub_rate_hz_).first;
+  this->dataPtr->pub_rate_ = std::make_unique<rclcpp::Rate>(this->dataPtr->pub_rate_hz_);
+
+  auto publish = [this]()
+    {
+      while (rclcpp::ok() && !this->dataPtr->stop_) {
+        if (this->dataPtr->tf_pub_->get_subscription_count() <= 0 && \
+          this->dataPtr->imu_pub_->get_subscription_count() <= 0) {continue;}
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:192]: `&& this->dataPtr->imu_pub_->get_subscription_count() <= 0`

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+          this->dataPtr->tf_record_.seq_num =
+            this->dataPtr->seq_num++ % std::numeric_limits<int16_t>::max();
+          tf_record = this->dataPtr->tf_record_;
+          data.unlock();
+          if (this->dataPtr->tf_pub_->get_subscription_count() > 0) {
+            this->dataPtr->tf_pub_->publish(tf_record);
+          }
+          if (this->dataPtr->imu_pub_->get_subscription_count() > 0) {
+            this->dataPtr->imu_pub_->publish(tf_record.imu);
+          }
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:212]: do similar for mag_pub_

**Comment by andermi** [comment]
Made some comments... also, not sure why the `no_inputs` test is timing out all 10 retries

**Comment by andermi** [comment]
oops @ `close with comment` 😆

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp (hunk 3 above)
Comment (quarkytale) [reply buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:64]: Should the `tf_record_.header`s be reset in `set_tf_record_mag`?

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+  std::unique_lock next(this->dataPtr->next_access_mutex_);
+  std::unique_lock data(this->dataPtr->data_mutex_);
+  next.unlock();
+
+  auto sec_nsec = gz::math::durationToSecNsec(this->dataPtr->current_time_);
+
+  this->dataPtr->tf_record_.header.stamp.sec = sec_nsec.first;
+  this->dataPtr->tf_record_.header.stamp.nanosec = sec_nsec.second;
+  //  Sea pressure: depth*rho*g, Pascal to psi
+  this->dataPtr->tf_record_.pressure = (depth * 1025 * 9.8)/6894.75729;
```
Comment (quarkytale) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:289]: Needs validation

Comment (andermi) [reply buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:289]: looks good!

**Comment by andermi** [comment]
LGTM after fixing uncrustify... also still not sure why the `no_inputs` test is failing in the PR CI

**Review by andermi — APPROVED** [review]

File: buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+#include "TrefoilController.hpp"
+
+struct buoy_gazebo::TrefoilControllerPrivate
+{
+  rclcpp::Node::SharedPtr rosnode_{nullptr};
+  rclcpp::executors::MultiThreadedExecutor::SharedPtr executor_;
+
+  gz::sim::Entity entity_;
+  gz::transport::Node node_;
+  bool use_sim_time_{true};
```
Comment (andermi) [inline buoy_gazebo/src/controllers/TrefoilController/TrefoilController.cpp:44]: Also need to set the `use_sim_time` parameter for `rosnode_` like in the `PowerController`

https://github.com/osrf/buoy_sim/blob/0b03fbda0550234850d1aed838b5a3433df38f97/buoy_gazebo/src/controllers/PowerController/PowerController.cpp#L215-L218

**Pull request merged**